// alumina/dist/cjs/core/functionComponentWrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFunctionComponentWrapperCached = void 0;
const hookImpl_1 = require("../hookImpl");
const promise = Promise.resolve();
function doLater(fn) {
    promise.then(fn);
}
function stringifyClasses(classes) {
    if (Array.isArray(classes)) {
        return classes.filter((a) => !!a).join(' ');
    }
    else {
        return classes;
    }
}
function createFunctionComponentWrapper(renderFunction) {
    const fcName = renderFunction.name;
    return {
        name: fcName,
        mount(self, props) {
            // console.log('mount', fcName);
            self.fcsig = fcName;
            self.hook = (0, hookImpl_1.createHookInstance)();
            self.renderWithHook = (props) => {
                (0, hookImpl_1.startHooks)(self.hook);
                const vnode = renderFunction(props);
                if (vnode) {
                    vnode.marker = `${fcName}`;
                }
                if (props.class &&
                    vnode &&
                    (vnode.vtype === 'vElement' || vnode.vtype === 'vComponent')) {
                    const fcClasses = stringifyClasses(props.class);
                    vnode.props.class = vnode.props.class
                        ? `${vnode.props.class} ${fcClasses}`
                        : fcClasses;
                }
                (0, hookImpl_1.endHooks)();
                doLater(() => (0, hookImpl_1.flushHookEffects)(self.hook));
                return vnode;
            };
            return self.renderWithHook(props);
        },
        update(self, props) {
            return self.renderWithHook(props);
        },
        unmount(self) {
            // console.log('unmount', fcName);
            (0, hookImpl_1.flushHookEffects)(self.hook, true);
        },
    };
}
function getFunctionComponentWrapperCached(renderFunction) {
    if (!renderFunction.__AluminaFunctionComponentWrapper) {
        renderFunction.__AluminaFunctionComponentWrapper =
            createFunctionComponentWrapper(renderFunction);
    }
    return renderFunction.__AluminaFunctionComponentWrapper;
}
exports.getFunctionComponentWrapperCached = getFunctionComponentWrapperCached;
