// alumina/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.render = exports.asyncRerender = exports.rerender = exports.effectOnMount = exports.useMemoWithArgs = exports.domStyled = exports.Fragment = exports.useContext = exports.createContext = exports.useRef = exports.useInlineEffect = exports.useEffect = exports.useCallback = exports.useMemo = exports.useLocal = exports.useState = exports.applyGlobalStyle = exports.styled = exports.css = exports.jsx = void 0;
require("./jsxTypes");
const aluminaGlobal_1 = require("./aluminaGlobal");
const contextApi_1 = require("./contextApi");
Object.defineProperty(exports, "createContext", { enumerable: true, get: function () { return contextApi_1.createContext; } });
Object.defineProperty(exports, "useContext", { enumerable: true, get: function () { return contextApi_1.useContext; } });
const core_1 = require("./core");
Object.defineProperty(exports, "Fragment", { enumerable: true, get: function () { return core_1.Fragment; } });
Object.defineProperty(exports, "jsx", { enumerable: true, get: function () { return core_1.jsx; } });
const cssInJs_1 = require("./cssInJs");
Object.defineProperty(exports, "applyGlobalStyle", { enumerable: true, get: function () { return cssInJs_1.applyGlobalStyle; } });
Object.defineProperty(exports, "css", { enumerable: true, get: function () { return cssInJs_1.css; } });
Object.defineProperty(exports, "styled", { enumerable: true, get: function () { return cssInJs_1.styled; } });
const hookImpl_1 = require("./hookImpl");
Object.defineProperty(exports, "useCallback", { enumerable: true, get: function () { return hookImpl_1.useCallback; } });
Object.defineProperty(exports, "useEffect", { enumerable: true, get: function () { return hookImpl_1.useEffect; } });
Object.defineProperty(exports, "useInlineEffect", { enumerable: true, get: function () { return hookImpl_1.useInlineEffect; } });
Object.defineProperty(exports, "useLocal", { enumerable: true, get: function () { return hookImpl_1.useLocal; } });
Object.defineProperty(exports, "useMemo", { enumerable: true, get: function () { return hookImpl_1.useMemo; } });
Object.defineProperty(exports, "useRef", { enumerable: true, get: function () { return hookImpl_1.useRef; } });
Object.defineProperty(exports, "useState", { enumerable: true, get: function () { return hookImpl_1.useState; } });
const extensions_1 = require("./extensions");
Object.defineProperty(exports, "domStyled", { enumerable: true, get: function () { return extensions_1.domStyled; } });
Object.defineProperty(exports, "effectOnMount", { enumerable: true, get: function () { return extensions_1.effectOnMount; } });
Object.defineProperty(exports, "useMemoWithArgs", { enumerable: true, get: function () { return extensions_1.useMemoWithArgs; } });
(0, cssInJs_1.setJsxCreateElementFunction)(core_1.jsx);
function rerender() {
    aluminaGlobal_1.aluminaGlobal.rerender();
}
exports.rerender = rerender;
function asyncRerender() {
    aluminaGlobal_1.aluminaGlobal.asyncRerenderFlag = true;
}
exports.asyncRerender = asyncRerender;
function setupAsyncRenderLoop() {
    if (!aluminaGlobal_1.aluminaGlobal.asyncLoopInitialized) {
        const asyncRenderLoop = () => {
            if (aluminaGlobal_1.aluminaGlobal.asyncRerenderFlag) {
                aluminaGlobal_1.aluminaGlobal.rerender();
                aluminaGlobal_1.aluminaGlobal.asyncRerenderFlag = false;
            }
            requestAnimationFrame(asyncRenderLoop);
        };
        asyncRenderLoop();
        aluminaGlobal_1.aluminaGlobal.asyncLoopInitialized = true;
    }
}
function render(renderFn, parentDomNode) {
    const executeRender = () => {
        (0, core_1.render)(renderFn(), parentDomNode);
        aluminaGlobal_1.aluminaGlobal.hookEffectFuncs.forEach((func) => func());
        aluminaGlobal_1.aluminaGlobal.hookEffectFuncs = [];
    };
    aluminaGlobal_1.aluminaGlobal.rerender = executeRender;
    executeRender();
    setupAsyncRenderLoop();
}
exports.render = render;
